### PR TITLE
feat: add blueprint methods

### PR DESCRIPTION
**Pull request by Yash-sudo-web**
Addresses Issue #79

### CODE REVIEW

**Comment by Yash-sudo-web** [comment]
hey @ryshu, how to proceed with this now, can you brief me a bit?

**Comment by Yash-sudo-web** [comment]
I updated the code according to the linter, can you re-run the tests and let me know if there are still any issues?

**Comment by ryshu** [comment]
Hi, you can run all pipelines tests on your own computer if you want.

Still missing all tests, pipelines will fail. 100÷ coverage is required to pass pipelines.

**Comment by Yash-sudo-web** [comment]
Hello, I have added the unittests for the blueprint methods, test coverage should be 100+ now.

**Comment by Yash-sudo-web** [comment]
hey @ryshu, can you review the PR? I am facing some issues with implementation of a unit test, one of them works fine but the other one is not passing.

File: novu/dto/__init__.py [hunk 1]
```diff
@@ -99,4 +100,6 @@
     "TenantDto",
     "TopicDto",
     "TriggerTopicDto",
+    "BlueprintDto",
```
Comment (ryshu) [inline novu/dto/__init__.py:103]: They are in alphabetical order here normally

File: novu/api/__init__.py [hunk 2]
```diff
@@ -34,4 +35,5 @@
     "SubscriberApi",
     "TenantApi",
     "TopicApi",
+    "BlueprintApi",
```
Comment (ryshu) [inline novu/api/__init__.py:38]: They are in alphabetical order here normally

File: tests/api/test_blueprint.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            blueprint_id="63dafeda7779f59258e38450",
+            created_at="2023-06-23T12:00:00.000Z",
+            updated_at="2023-06-23T12:00:00.000Z",
+            deleted=False,
+            deleted_at=None,
+            deleted_by=None,
+        )
+        cls.expected_grouped_blueprint_dto = GroupedBlueprintDto(
+        category="example_category",
+        general=[{}],
```
Comment (ryshu) [inline tests/api/test_blueprint.py:52]: API return and expectation aren't the same here. You expect and empty general when the API return blueprint DTO

File: novu/dto/blueprint.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    """List of triggers in the blueprint."""
+
+@dataclasses.dataclass
+class GroupedBlueprintDto(CamelCaseDto["GroupedBlueprintDto"]):
+    """Grouped blueprint definition."""
+
+    category: str
+    """Blueprint category."""
+
+    general: List[BlueprintDto] = dataclasses.field(default_factory=list)
```
Comment (ryshu) [inline novu/dto/blueprint.py:89]: This is not the correct way to code a list field containing a DataClass. You have to look at the other DTOs to understand how it is used (in the code, look at `DtoIterableDescriptor`)

File: novu/dto/blueprint.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    """List of general blueprints in the category."""
+
+    popular: List[BlueprintDto] = dataclasses.field(default_factory=list)
```
Comment (ryshu) [inline novu/dto/blueprint.py:92]: Same here

**Review by ryshu — CHANGES_REQUESTED** [review]
When I look at the history, I see that the commits you make don't follow the naming convention.

I also observe that the merge request is targeted towards the "main" branch while we use the "alpha" branch for current developments.

Since all of these points are covered in the contribution guidelines, I would like you to take a look at them before asking for my help. https://github.com/novuhq/novu-python/blob/main/CONTRIBUTING.rst

For the unit test part, the unit test fails and shows the error you are making. Since it is enough to read the result to understand the error, I don't see where I can be useful.

image

Result doesn't match your expectation.

FYI, for debug purpose only (should never be comited), you can use `self.maxDiff = None` before the test to see the untruncated message. https://docs.python.org/3/library/unittest.html#unittest.TestCase.maxDiff

File: novu/dto/__init__.py (hunk 1 above)
Comment (Yash-sudo-web) [reply novu/dto/__init__.py:103]: Done

File: novu/api/__init__.py (hunk 2 above)
Comment (Yash-sudo-web) [reply novu/api/__init__.py:38]: Done

File: novu/dto/blueprint.py (hunk 4 above)
Comment (Yash-sudo-web) [reply novu/dto/blueprint.py:89]: Done

File: novu/dto/blueprint.py (hunk 5 above)
Comment (Yash-sudo-web) [reply novu/dto/blueprint.py:92]: Done

**Comment by ryshu** [comment]
@Yash-sudo-web Do you want me to finish your MR? The unit tests don't seem to be too complex for me to solve and we can plan a new alpha quickly.

**Comment by Yash-sudo-web** [comment]
hey, @ryshu yes you can finish the MR, thanks.

**Review by ryshu — APPROVED** [review]

**Comment by ryshu** [comment]
@Cliftonz Hi, also ready for rebase and merge

**Pull request merged**